### PR TITLE
add a boxed function for conditional functionalization

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR is a POC for conditional functionalization, tested with `diag_embed`

Test:
```
>>> import torch
>>> from  functorch import vmap
>>> vmap(lambda x: torch.diag_embed(x))(torch.ones(2, 2, 2))
__main__:1: UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::diagonal_scatter. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at  /raid/hirsheybar/pytorch-functionalization/functorch/functorch/csrc/BatchedFallback.cpp:106.)
tensor([[[[1., 0.],
          [0., 1.]],

         [[1., 0.],
          [0., 1.]]],


        [[[1., 0.],
          [0., 1.]],

         [[1., 0.],
          [0., 1.]]]])
```

Note the warning: `diag_embed` decomposes into a call to `diag` (a view op), and to properly remove aliases the functionalization pass eventually makes a call to `diag_scatter` (a new op that I added in my functionalization pass PR stack). It should be easy enough to add batching rules for the new `diag_scatter`/`slice_scatter`/`select_scatter` ops that I added.

A couple things to note:

(1) This is the slightly more efficient version of conditional functionalization that "fuses" the functionalization and vmap passes into a single level. Basically, we set up TLS in such a way that when you call into `vmap(diag_embed)`, we sandwich `FunctionalizationKernel` -> `BatchingRule` between the FrontFallback and BackFallback kernels.

So the full dispatch tree ends up looking something like:

```
.  .  .  .  .  .  .  .  FrontFallback(composite) .  .  .  .  .  .  <--- adds [BatchedKey, BackModeKey] to TLS
.  .  .  .  .  .  .  .  .  .  .  .  |
.  .  .  .  .  Functional_Decompose_Kernel(composite) .  .  .  .  .  .  <--- adds [FunctionalKey] to TLS, hits at::native:: kernel
.  .  .  .  .  | .  .  .  .  .  .  .  . .  .  .  |
Functionalkernel(base1) .  .  .  .  .  .  Functionalkernel(base2)
.  .  .  .  .  | .  .  .  .  .  .  .  .  .  .   |
BatchingRule(base1).  .  .  .  .  .  BatchingRule(base2)
.  .  .  .  .   | .  .  .  .  .  .  .  .  .  .  |
BackFallback(base1).  .  .  .  .  .   BackFallback(base2)
```

(2) Annoyingly, I think the most common inplace op that composite kernels use (that breaks functorch) is `copy_`, which doesn't have a corresponding out-of-place variant. Technically, `copy_` would need an out-of-place **and** an out= variant in order to be supported. The above test only works locally because I added a hack in the codegen to deal with copy_.

I think that the out-of-place version of `copy_()` is pretty much just a call to `clone()`, so I burned "copy_() -> clone()" into the codegen locally to get things to work.

Maybe, if copy_() is an important enough one-off case, then this a justifiable hack? Otherwise we'd need to consider adding the other variants of `copy_`.

(3) I wrote it as a boxed kernel, which adds a bit of overhead (a second trip through the dispatcher to hit the at::native:: composite kernel). There's probably a messy template metaprogramming way to not used boxed fallbacks

### CODE REVIEW

File: functorch/csrc/BatchRulesStopDecomposition.cpp [hunk 1]
```diff
@@ -6,13 +6,98 @@
...
+      auto functional_ivalue = at::functionalization::impl::to_functional_tensor(ivalue.toTensor());
+      (*stack)[arguments_begin + idx] = std::move(functional_ivalue);
+    } else if (ivalue.isTensorList()) {
+      auto functional_ivalue = at::functionalization::impl::to_functional_tensor(ivalue.toTensorList());
+      (*stack)[arguments_begin + idx] = std::move(functional_ivalue);
+    }
+  }
+
+  // Step 2: set up TLS such that we hit the functionalization kernels before the batching rules.
+  // Note: this relies on the fact that Functionalization > BatchMode in DispatchKey.h
```
Comment (zou3519) [inline functorch/csrc/BatchRulesStopDecomposition.cpp:59]: I think I understand the problem you were mentioning last week where it is unclear how to "push the functionalize transform onto the dynamic layer stack" -- when we get to DECOMPOSE_FUNCTIONAL, the dynamic layer stack has already decided to dispatch to the vmap subsystem!

File: functorch/csrc/BatchRulesStopDecomposition.cpp [hunk 2]
```diff
@@ -624,7 +709,7 @@ TORCH_LIBRARY_IMPL(aten, FT_BATCHED_KEY, m) {
   STOP_DECOMPOSE(zeros.out);
 
   // These throw an error in our tests if we remove them
-  STOP_DECOMPOSE(diag_embed);
+  DECOMPOSE_FUNCTIONAL(diag_embed);
```
Comment (Chillee) [inline functorch/csrc/BatchRulesStopDecomposition.cpp:712]: Very awesome :)

**Comment by bdhirsh** [comment]
> How important is is that the result of functionalization maintains the same strides as the result without functionalization? If we have `x.copy_(y)`, the result (x) has the same strides as the input (x). With `y.clone()`, the strides are like y but not like x.

> `copy_` also has some broadcasting semantics in that `y` can be broadcast to the shape of `x`. So if we didn't care about the strides being preserved we could replace `x.copy_(y)` with `y.clone().expand_as(x)` (or even `y.expand_as(x)`! the .clone() is unnecessary if there is no mutation)

This seems important and not something that I tested 😛 . In particular, a programming that happened to call `x.copy_(y[0, :])` would probably just return incorrect results after getting functionalized (same thing with the broadcasting semantics you pointed out).

There's also actually a third problem with my `clone_()` -> `copy()` burn-in in the the codegen, that doesn't get uncovered from the diag_embed example. Which is that if I were to call:
```
batched_tensor.copy_(cpu_tensor)
```
`batched_tensor` will (correctly) continue to be a batched tensor after the inplace op. How does that happen? `copy_()` takes in two arguments, and in this case one of them is a batched tensor and one isn't. That causes us functorch to dispatch to the batching rule, and ultimately return another batched tensor.

We want the functionalized version to do the same, but instead we just end up with a call to `cpu_tensor.clone()`, which skips the vmap() transform in the layer stack completely and returns another cpu tensor.

So I'm kind of convinced at this point that we need an out-of-place variant of `copy_()` that has its own batching rule (or not, since it could hit the for loop fallback).

Technically in the current form of the codegen, I'd also need a `copy.out` variant, which would force the codegen to group the 3 variants into a `NativeFunctionsGroup` and pick up the mapping from inplace->out_of_place automatically. I could probably avoid that by just manually making my own grouping in the functionalization codegen, between inplace + out_of_place ops.

File: functorch/csrc/BatchRulesStopDecomposition.cpp [hunk 3]
```diff
@@ -6,13 +6,98 @@
...
 #include <functorch/csrc/BatchRulesHelper.h>
+#include <ATen/FunctionalTensorWrapper.h>
 #include <ATen/Operators.h>
 #include <functorch/csrc/PlumbingHelper.h>
 #include <functorch/csrc/BatchedFallback.h>
 #include <ATen/core/dispatch/Dispatcher.h>
 
 namespace at { namespace functorch {
 
+at::Tensor sync_and_unwrap_functional_output(at::Tensor out_functional) {
```
Comment (zou3519) [inline functorch/csrc/BatchRulesStopDecomposition.cpp:17]: NB: we deleted this file but you can probably just plop the code into functorch/csrc/BatchRulesDecompositions.cpp

**Comment by zou3519** [comment]
I'm cleaning up old PRs in preparation for the functorch->pytorch merge; we've already landed functionalization in some form so closing this.

**Comment by bdhirsh** [comment]
sorry for not cleaning these up :)

**Comment by zou3519** [comment]
No need to apologize, I leave PRs open all the time too haha

**Pull request closed without merging**